### DISCUSSION TITLE
understanding license / block coding - personal use w/ the kids, pc & tablet

### DISCUSSION THREAD

**Discussion by m10d** [pybricks/support#1665] [Q&A]
Hi - interested in jumping to pybricks now that block coding is supported.  Don't mind paying $60 one time to support the project.

I can't find answers to some basic questions, thought:

- can one use the tablet as a live controller?  IE  add elements to your program to that result in accepting input from the tablet.  similar _functionality_ to how you might make an "RC car" with say buwizz & buwizz app; presumably this would still run code autonomously on the hub (as in my "confirm please" below), but the device would be talking over something like HID or bluetooth uart to the tablet?  (and as a consequence of being able to control the code on the fw side (contrary to buwizz & similar) one still has full use of sensors, etc.
- the kids program on both a tablet and a PC depending on what their doing (code is synced via syncthing, and **only one** is used at one time (but both may be used several times in a day) - is this supported with the personal perpetual license?

and just for confirmation (no response to these needed unless I read it wrong):

- just to confirm: all features of the hubs run in "autonomous" mode, still, with ~scratch type programming, correct?  The tablet/pc/phone is only used to code & send code/firmware - bluetooth can be off entirely during operation?
- also just to confirm: looks like the programs on the hub do NOT support gamepads directly (presumably a limitation w/ the hub? - I see the ESP32 addon which presumably talks serial (but this must consume one of the PU ports :(  which is a big limiter for use of the technic hubs as we use all the ports) - 

Thank you!

**Comment by laurensvalk** [comment]
Thanks for your interest in the Pybricks project!

> can one use the tablet as a live controller? 

You can use the [keyboard as input](https://pybricks.com/project/pybricks-3-5-released/#new-block-for-keyboard-input), though this isn't as fancy as interactive on-screen widget. I can recommend the LEGO Powered Up Remote though (about $15) or the Xbox Controller ($50), which feels much nicer than a touch screen.

> the kids program on both a tablet and a PC depending on what their doing

Completely fine. We try to make it as easy and unintrusive as possible :slightly_smiling_face: 

I'll respond in more detail to your [other question](https://github.com/orgs/pybricks/discussions/277#discussioncomment-9679525).

>  all features of the hubs run in "autonomous" mode, still, with ~scratch type programming, correct? The tablet/pc/phone is only used to code & send code/firmware

Correct. The blocks generate Python code (which you can also see happen live as you move blocks around). The robot doesn't see the difference.

We've taken care to generate "nicely looking" code by carefully designing the blocks to match the underlying Python API, so there is very little filler or glue code that slows things down as it would with e.g. the official apps. 

> bluetooth can be off entirely during operation?

Once you've uploaded your program - yes. The program will still be there when you turn the hub off and on again.

> looks like the programs on the hub do NOT support gamepads directly

They do. The Xbox controller is supported with the Technic / Spike / Mindstorms hubs. [Here are a few example projects.](https://pybricks.com/project/xbox-controller/)

**Reply by m10d** [reply]
> > can one use the tablet as a live controller?
> 
> You can use the [keyboard as input](https://pybricks.com/project/pybricks-3-5-released/#new-block-for-keyboard-input), though this isn't as fancy as interactive on-screen widget. I can recommend the LEGO Powered Up Remote though (about $15) or the Xbox Controller ($50), which feels much nicer than a touch screen.

That's fantastic!  

- On the keyboard, I have searched by can't find docs beyond https://pybricks.com/projects/tutorials/wireless/hub-to-device/pc-keyboard/ - how do you actually connect a keyboard?  Under the hood is this just a bluetooth HID device, and the limitations are just that the pybricks code only expects/accepts the ~ascii input?  
- controller: Out of curiosity, it sounds plausible that's the same limitation for supporting just the xbox bluetooth controllers, that the code only has mapping for those controls?  That's all just fine, I have 8bito, but sounds like I'll be getting the xbox version - this is a massive improvement on status quo (especially when we're talking about young kids - the target of block-level 'languages' - to be able to **directly** use a normal ~"RC" / game controller for live play vs requiring mom/dad's phone to be in the loop!

**Reply by laurensvalk** [reply]
> > > can one use the tablet as a live controller?
> >
> > You can use the [keyboard as input](https://pybricks.com/project/pybricks-3-5-released/#new-block-for-keyboard-input),
>
> how do you actually connect a keyboard? 

This was in response to using your tablet/pc as a live controller. The input/output pane in the Pybricks app can be used for that when you are connected.

> he same limitation for supporting just the xbox bluetooth controllers, that the code only has mapping for those controls?

The Xbox Controllers use Bluetooth Low Energy, which is what the hubs also use. If another gamepad supports this, then it could theoretically be added to Pybricks, but we don't have any other controllers right now.

> this is a massive improvement on status quo (especially when we're talking about young kids - the target of block-level 'languages' - to be able to directly use a normal ~"RC" / game controller for live play vs requiring mom/dad's phone to be in the loop!

Yep!
